### PR TITLE
[BIOMAGE-1583] - Fix source file download

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1583

#### Link to staging deployment URL 

https://ui-agi-s3-download.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

UI : https://github.com/biomage-ltd/ui/pull/573
API : https://github.com/biomage-ltd/api/pull/266

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Code updates
Have best practices and ongoing refactors being observed in this PR

- [x] Migrated any selector / reducer used to the new format
- [x] Validated that current unit tests work as expected and are enough

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [x] Deployed to staging
- [x] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/redux/actions/samples/updateSampleFile.js [hunk 1]
```diff
@@ -32,6 +32,10 @@ const updateSampleFile = (
       };
 
       const newSample = mergeObjectWithArrays(sample, diffObject);
+
+      // Do not save the fileObject to DynamoDB
+      delete newSample.files[fileName].fileObject;
```
Comment (cosa65) [inline src/redux/actions/samples/updateSampleFile.js]: shouldn't this delete be done in `saveSamples`? Since that's where it is being saved to dynamo, because if `saveSamples` is run outside of this `updateSampleFIle` we stil have the problem, right?

File: src/components/data-management/UploadDetailsModal.jsx [hunk 2]
```diff
@@ -39,7 +36,7 @@ const UploadDetailsModal = (props) => {
 
   useEffect(() => {
     if (replacementFileBundle) {
-      bundleToFile(replacementFileBundle, SELECTED_TECH).then((newFile) => {
+      fileObjectToFileRecord(replacementFileBundle, SELECTED_TECH).then((newFile) => {
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: You have the "File bundle" syntax you are changing in other places too, like `uploadFileBundle`

Comment (cosa65) [reply src/components/data-management/UploadDetailsModal.jsx]: If part of what you are doing in this ticket is renaming file bundle to file object I would change it everywhere

File: src/utils/data-management/downloadSingleFile.js [hunk 3]
```diff
@@ -1,14 +1,9 @@
-import { saveAs } from 'file-saver';
-import { Storage } from 'aws-amplify';
+import downloadFromUrl from 'utils/data-management/downloadFromUrl';
+import { getFromApiExpectOK } from 'utils/getDataExpectOK';
 
-const downloadSingleFile = async (activeProjectUuid, sampleUuid, fileName, bundleName) => {
```
Comment (cosa65) [inline src/utils/data-management/downloadSingleFile.js:4]: We were using amplify? And it was failing because the permissions to access this from amplify had already been removed?

Comment (aerlaut) [reply src/utils/data-management/downloadSingleFile.js:4]: Yup

File: src/components/data-management/UploadDetailsModal.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-management/UploadDetailsModal.jsx]: replaced

File: src/redux/actions/samples/updateSampleFile.js (hunk 1 above)
Comment (aerlaut) [reply src/redux/actions/samples/updateSampleFile.js]: Hmm... right, that's true. I'll move this delete inside `saveSamples`.

Comment (aerlaut) [reply src/redux/actions/samples/updateSampleFile.js]: This call is moved inside `saveSamples`. Please refer to the new changes by looking at the file from the "Files changed" tab.

File: src/components/EditableParagraph.jsx [hunk 4]
```diff
@@ -10,6 +10,7 @@ const EditablePagrapraph = (props) => {
 
   const paragraphEditor = useRef();
 
+  const [text, setText] = useState(value);
```
Comment (cosa65) [inline src/components/EditableParagraph.jsx:13]: What are the changes done here for? They don't seem related to the ticket

Comment (cosa65) [inline src/components/EditableParagraph.jsx:13]: It seems the `text` `useState` is being used as a local version of `value`, so why not call it `localValue` or something like that to make that clearer? It is better to use one cohesive name.

If you feel `value` is not good I would change `value` to `text` and `text` to `localText`,

File: src/utils/upload/processUpload.js [hunk 5]
```diff
@@ -217,16 +216,16 @@ const processUpload = async (filesList, sampleType, samples, activeProjectUuid,
   });
 };
 
-const bundleToFile = async (bundle, technology) => {
+const fileObjectToFileRecord = async (fileObject, technology) => {
```
Comment (cosa65) [inline src/utils/upload/processUpload.js:234]: `bundleToFile` wasn't very self explanatory, but `fileObjectToFileRecord` isn't very explanatory either, what do you think of perhaps coming up with a better name or adding comments explaining what is the input and what is the output?

Example of comments.

  ```
/**
   * Does cool things.
   * @param {string} fileObject blalbalba.
   * @param {string} technology blsblslfs.
   * @returns {object} fileRecord blablabladblsal
   */
   
```

File: src/components/EditableParagraph.jsx (hunk 4 above)
Comment (cosa65) [reply src/components/EditableParagraph.jsx:13]: Ah you added unit tests so codecov doesn't cry I think, but the refactoring here was to make this respond better or something?

Comment (aerlaut) [reply src/components/EditableParagraph.jsx:13]: So that it can have internal value, and the component can be independently without outside values

Comment (aerlaut) [reply src/components/EditableParagraph.jsx:13]: I think this is already pretty clear. We usually pass in `value` to an input component. However, inside the component I used `text` because I'm not really using a real `input` component, but a div with `contenteditable`, which should display some kind of `text` inside it.

File: src/utils/upload/processUpload.js (hunk 5 above)
Comment (aerlaut) [reply src/utils/upload/processUpload.js:234]: I've added a comment to explain the function,

File: src/components/EditableParagraph.jsx (hunk 4 above)
Comment (cosa65) [reply src/components/EditableParagraph.jsx:13]: But `text` is the same as `value` but internal, so why use a completely different name?

I think it is clear right now because there isn't much code, the problem is more noticeable in larger files but using cohesive names is still a good practice.

Other naming ideas could be `inputText` `text` or `inputValue` and `value`, you can rename those on destructuring  the `props` so that from outside we only see `value` if you want.

Comment (aerlaut) [reply src/components/EditableParagraph.jsx:13]: Hmm... ok, that's a good argument. I'll change the name,

File: src/components/data-management/UploadDetailsModal.jsx [hunk 6]
```diff
@@ -25,30 +25,27 @@ const UploadDetailsModal = (props) => {
   } = props;
   const { fileCategory, sampleUuid } = uploadDetailsModalDataRef.current || false;
   const file = uploadDetailsModalDataRef.current?.file || {};
-  const {
-    upload = {}, bundle = {},
-  } = file;
+  const { upload } = file || {};
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: Use `??` instead for null checks

Comment (cosa65) [reply src/components/data-management/UploadDetailsModal.jsx]: also perhaps do this same change to 
```
const { fileCategory, sampleUuid } = uploadDetailsModalDataRef.current || false;
  const file = uploadDetailsModalDataRef.current?.file || {};
```

Comment (cosa65) [reply src/components/data-management/UploadDetailsModal.jsx]: Also if you can change the usage of false to null instead in `uploadDetailsModalDataRef.current || false`

File: src/redux/actions/samples/saveSamples.js [hunk 7]
```diff
@@ -15,21 +16,29 @@ const saveSamples = (
...
   // add new sample to payload
   if (addSample) {
     const projectSamples = projects[projectUuid].samples.reduce((acc, sampleId) => {
       acc[sampleId] = samples[sampleId];
       return acc;
     }, {});
 
+    // Do not save the fileObject to DynamoDB as it can not be serialized
+    Object.keys(newSampleCopy.files).forEach((file) => {
+      delete newSampleCopy.files[file].fileObject;
```
Comment (cosa65) [inline src/redux/actions/samples/saveSamples.js]: We decided to disable retry button when the file is fetched from dynamo (because then we don't have fileObject to perform the retry).

File: src/redux/actions/samples/saveSamples.js [hunk 8]
```diff
@@ -15,21 +16,29 @@ const saveSamples = (
...
+    });
+
     payload = projectSamples;
     payload = {
       ...payload,
-      [newSample.uuid]: newSample,
+      [newSample.uuid]: newSampleCopy,
```
Comment (cosa65) [inline src/redux/actions/samples/saveSamples.js]: rename to `newSampleToUpload`

File: src/utils/upload/processUpload.js [hunk 9]
```diff
@@ -66,13 +66,20 @@ const compressAndUploadSingleFile = async (
...
+  };
+
   try {
     loadedFile = await loadAndCompressIfNecessary(file, () => (
       dispatch(
         updateSampleFile(
           sampleUuid,
           fileName,
-          { upload: { status: UploadStatus.COMPRESSING } },
+          {
```
Comment (cosa65) [inline src/utils/upload/processUpload.js]: super nitpick: compress again this and other line

File: src/components/data-management/UploadDetailsModal.jsx (hunk 6 above)
Comment (aerlaut) [reply src/components/data-management/UploadDetailsModal.jsx]: changed

File: src/redux/actions/samples/saveSamples.js (hunk 8 above)
Comment (aerlaut) [reply src/redux/actions/samples/saveSamples.js]: changed

File: src/utils/upload/processUpload.js (hunk 9 above)
Comment (aerlaut) [reply src/utils/upload/processUpload.js]: changed

File: src/redux/actions/samples/saveSamples.js (hunk 7 above)
Comment (aerlaut) [reply src/redux/actions/samples/saveSamples.js]: added

**Review by cosa65 — APPROVED** [review]

**Review by vickymorrison — APPROVED** [review]
